doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,266 @@
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
+Solution
+--------
+
+In OpenSSL most of the existing code in the FIPS provider is using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+typedef int (OSSL_CALLBACK)(const OSSL_PARAM params[], void *arg);
+
+void OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+The callback and/or cbarg can be changed at any time.
+
+- Internally within the FIPS module algorithms the following internal helper
+callback may be called
+
+```c
+int ossl_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
+                            const char *algdesc)
+```
+
+An application's indicator OSSL_CALLBACK can be used to log that an
+indicator was triggered. The callback may return either zero or non zero based
+on the indicator type and description. Returning non zero from the callback
+allows the operation to continue in a non-FIPS approved mode of operation.
```

Comment:
Is it even ok for a FIPS module to change behavior in this way, during an operation, based on external inputs ?

@jvdsn isn't there a chance this could be seen as a bypass capability (7.4.3.2) ?

Reply:
It's possible. Bypass capability mainly makes the distinction between cryptographic protection or no protection. However, non-approved aren't considered to provide security in FIPS so maybe from that perspective there wouldn't be protection on the output. Alternatively it might be possible to argue that they're two different services, and the callback selects which service to use. But I'd say that depends on when the callback is called inside the service.

From the code below, it looks like the strict_checks config option would override the callback system, so it's probably easiest to simply ensure the config always has strict checks enabled.

Reply:
The main intention of the callback is to provide error logging when non approved mode is being used. The  intent of the return value from the callback was mainly for testing purposes. (Since we need to test all these paths).

Reply:
Currently the callback is only triggered when it is already non approved. So returning an error from here doesnt seem like an issue.

Reply:
> Currently the callback is only triggered when it is already non approved. So returning an error from here doesnt seem like an issue.

In that case I don't think there should be concern about the bypass. In what sense is the return value required for testing purposes? Wouldn't it be sufficient to test everything once with strict_checks and then once without strict_checks?

Reply:
There is a 'strict' setter and getter for each algorithm ctx, as well as the config setting - so there is 3 combinations. We can make the callback fail for testing purposes this way (as I have done). I think it is useful to have the ability to cause a failure so that it doesnt silently fail (Obviusly it is not completely silent if it logs - but testing for logs is a bit harder).

Reply:
From a behavioral standpoint, would there be a difference between:
- Setting `EVP_default_properties_enable_fips(NULL, 1)` with `strict = false` in the config versus setting `EVP_default_properties_enable_fips(NULL, 0)`
- Performing `EVP_*_fetch(NULL, "...", "fips=yes")` with `strict = false` in the algorithm context?

Reply:
I anticipate that there are going to be checks for NIST madness that aren't covered by the more global strict setting.  We could still go either way on this.

There is a world of different between `EVP_*_fetch(NULL, "md5", "fips=yes")` with strict = false in the algorithm context, versus `EVP_*_fetch(NULL, "md5", "fips=no")`.

Likewise for `EVP_*_fetch(NULL, "sha512", "fips=yes")` with strict = false in the algorithm context, versus `EVP_*_fetch(NULL, "sha512", "fips=no")`, the former guarantees the FIPS implementation, the latter doesn't.  Some agencies need to use validated FIPS implementations but in a way that isn't supported by the standards anymore (legacy use predominately).

Reply:
@slontis if error logging is what you care for, why not use the regular error logging facility you already have?

You set a parameter of "logging = 1" and if that is set you use the normal error logging log in there each check that failed, and finally return an error from the function with the error stack intact.

The application can then just pull the regular error log and print it or examine it as needed.